Replace most usages of context.TODO()

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
`context.TODO()` is currently used in quite a few places instead of a proper context. This delays interruption of restic in some cases. In most places we already have a proper context at hand or nearby, so use these.

The only remaining usages are in `cmd/restic/lock.go` and `internal/restic/lock.go` to ensure that lock refresh and cleanup still work while the rest of restic prepares for shutdown. And for `internal/restic/readerat.go` there's no clean way to pass down a proper context.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------
This fixes the remaining parts to resolve #1434.

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] **I don't think that there are context cancellation tests in most places. However, the existing tests should at least ensure that the context doesn't break normal executions** I have added tests for all changes in this PR
- ~~[ ] I have added documentation for the changes (in the manual)~~
- [ ] **What someone might notice is that some commands could react a bit faster to being interrupted** There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

MichaelEischer committed:
pass context to Find / FindSnapshot

This allows proper interruption of restic while it searches for
snapshots or key files.

MichaelEischer committed:
debug: use proper context

This allows the debug commands to be properly interrupted.

MichaelEischer committed:
restore: Use proper context while loading snapshot

MichaelEischer committed:
init: pass proper context to master key generation

This is no change in behavior as a canceled context did later on cause
the config file creation to fail. Therefore this change just lets the
repository initialization fail a bit earlier.

MichaelEischer committed:
integration_fuse_test: use global context

No need to use the TODO context.

MichaelEischer committed:
restorer: remove redundant type specification

MichaelEischer committed:
unlock: use proper context for locks cleanup

The list operation used by RemoveStaleLocks or RemoveAllLocks will
already be canceled by the passed in context. Therefore we can also just
cancel the remove operation as the unlock command won't process all lock
files anyways.

MichaelEischer committed:
rest: use global context on repository creation

MichaelEischer committed:
Pass context to lockRepo

MichaelEischer committed:
Skip unlocking for nil locks

Now that lockRepo receives a context, it is possible that it is canceled
before a lock was created. Thus `unlockRepo` must be able to handle this
case.

MichaelEischer committed:
pass proper context into MasterIndex.RebuildIndex

greatroar commented:
LGTM. For backendReaderAt, we could just pass a Context to the constructor, store it and explain in the docs that the ReaderAt should not be reused across contexts. It's ugly, but I've done it before in custom Readers.

MichaelEischer committed:
Embed context into ReaderAt

The io.Reader interface does not support contexts, such that it is
necessary to embed the context into the backendReaderAt struct. This has
the problem that a reader might suddenly stop working when it's
contained context is canceled. However, this is now problem here as the
reader instances never escape the calling function.

MichaelEischer commented:
> For backendReaderAt, we could just pass a Context to the constructor, store it and explain in the docs that the ReaderAt should not be reused across contexts. It's ugly, but I've done it before in custom Readers.

I've embedded the context and added a comment that the created reader should not escape the calling function.

fd0 reviewed (approved):
Awesome, thanks! LGTM!
